97.md

```diff
@@ -0,0 +1,251 @@
+
+Key Takeaways
+-------------
+
+* Defines how binary content (images, media, files) is managed by NOSTR network
+* Allows sending private binary content in direct messages (encrypted)
+* Defines new commands at the protocol level ("ATTACH","FETCH")
+* Defines a new tag "attachment"
+* Uses the binary messages in the websocket connection
+
+
+New tag "attachment"
+-------------
+The proposal introduces a new tag  "attachment" with the following format
+
+```
+["attachment","hash_hex","size_in_bytes","mime_type","features1", "feature2", ....]
+```
+
+* **hash_hex** - hexadecimal representation of SHA256 hash of the attached binary file. This string also serves as **attachment-id**
+* **size_in_bytes** - size of the attachment in bytes
+* **mime_type** - mime type of the attachment 
+* **features...** - optional fields with various features in format `<field>=<value>`
+    * **blurhash=** - contains NIP-94 blurhash
+    * **purpose=** - purpose of the attachment in context of current note. This can help client
+to select optimal rendering. Suggested purposes
+        * **thumbnail** - thubnail image of a following video
+        * **story** - all attachments marked by this purpose are played as a story
+        * **inline** - attachment is rendered in text, referenced by `#[<index>]`. This purpose hides this attachment from the list of attachments in note's detail.
+
+```
+["attachment","12887...","64000","image/jpeg","blurhash=qwdjq3...", "purpose=thumbnail"]
+["attachment","874788...","2258147","video/mpeg4", "purpose=story"]
+["attachment","ae3c758...","2158799","video/mpeg4", "purpose=story"]
+["attachment","788eq78...","254123","image/jpeg", "purpose=story"]
+```
+        
+
+* One event can have more than one of these tags, then it carries more attachments.
+* Each relay defines the maximum attachment size and also the maximum number of attachments within an event
+* Attachment can be attached to any "kind". Attachments for `kind:4` are always encrypted
+
+The presence of the `attachment` attribute only informs the client that the relay should also have its own attachment content available in binary form. The client must download the attachments from the same relay to display them. See below
+
+
+Upload
+------
+Single event must be published simultaneously with attachments. It is necessary to prevent the client from publishing the event but not uploading the relevant attachments. Similarly, it should not be possible to upload attachments first and then publish the event that references them. The operation must be atomic
+
+For this reason, a new command is introduced: "ATTACH"
+
+**Protocol flow**
+
+```
+client: ["ATTACH", <event>]
+relay:  ["OK","<event-id">,true,"continue"]
+client: <binary message for first attachment>
+relay:  ["ATTACH",<attachment-id>, true, "continue"]
+client: <binary message for second attachment>
```

Comment:
Add the attachment-id to the binary content to avoid race conditions in multi-threaded, multi-connection environments. Nostr doesn't have a notion of atomic transaction for a sequence of messages and there is no need to create one for this.

Reply:
No one here is talking about a multithreading environment. 
The word atomic is meant in the sense of database transactions.
It means that in one state the relay does not have an event and attachments and in the other state it has the event and attachments at the same time. There should not be a situation where the relay will have an event without attachments or attachments without the event (in the latter case, it is more of a problem with the GC)

Reply:
> There should not be a situation where the relay will have an event without attachments or attachments without the event. 

But there are.... In 2 particular ways: 

1. All relays that do not implement this NIP will have events without attachments and there is nothing you can do about it. You might not care about it, but I HAVE to. The client always has to deal with multiple relays out there. Even the crazy ones. 

2. Deleting events happen all the time these days. Deleting attachment tags from a Replaceable events is also common. Either way, the GC is going to take its time. Meanwhile, for clients, your relay is offering attachments without events. As a client, I must ALWAYS expect that the relay is broken, because the GC has run yet. 

Because of 1 and 2, your atomic lock doesn't work. 

There is no need to make guarantees you cannot keep as a relay. If you can't absolutely guarantee them, there is no need to have them in the first place. 

In other words: It's ok if you have orphan attachments. You don't need to make a solution for that.